parsing/lexer.mll

```diff
@@ -104,6 +105,76 @@ let keyword_table =
     "asr", INFIXOP4("asr")
 ]
 
+let refresh_keywords tbl =
+  Hashtbl.clear tbl;
+  Hashtbl.iter
+    (fun k v -> Hashtbl.replace tbl k (Some v))
+    keyword_table_all
+
+let keyword_table =
+  let tbl = Hashtbl.create 149 in
+  refresh_keywords tbl;
+  tbl
+
+let keyword_edition v =
+  Option.iter (fun v ->
+    if v < (5, 3) then
+      Hashtbl.remove keyword_table "effect"
+  ) v
+
+let parse_keyword_edition s =
```

Comment:
Not fond of the overall logic design here. The Lexer is a busy enough place, with a lot of subtle stuff, and now we are adding configuration parsing and interpretation inside it. Here would be my preferred API:

- when we initialize the lexer, we can pass it a list or set of keywords
  (this could be an argument to `Lexer.init`, or an alternate initializer for compatibility-reason)
- we have code *somewhere else* in the compiler codebase, for example in our beloved utils/misc.ml module (a submodule Keyword_edition?) that parses the configured edition and turns it into a list or set of supported keywords

If I understand correctly, the logic in the lexer initialization code would then be to populate its keyword table from an immutable keyword->token association list and an (optional) keyword list: for each keyword in the keyword list, add them to the keyword table, with None if they are missing from the association list. This is much less code in lexer.mll than you currently have, and I like it.

Reply:
The flag parsing logic could certainly be moved to `Clflags`, and having the list of keywords as an optional argument of `Lexer.init` would work too.